nx/lib/nx/testing.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Nx.Testing do
+  @moduledoc """
+  Testing functions for Nx tensor assertions.
+
+  This module provides functions for asserting tensor equality and
+  approximate equality within specified tolerances.
+  """
+
+  import ExUnit.Assertions
+  import Nx, only: [is_tensor: 1]
+
+  @doc """
+  Asserts that two tensors are exactly equal.
+
+  This handles NaN values correctly by considering NaN == NaN as true.
+  """
+  def assert_equal(left, right) when not is_tensor(left) or not is_tensor(right) do
+    if not Nx.Defn.Composite.compatible?(left, right, &tensor_equal?/2) do
+      flunk("""
+      Tensor assertion failed.
+      left: #{inspect(left)}
+      right: #{inspect(right)}
+      """)
+    end
+  end
+
+  def assert_equal(left, right) do
+    if !tensor_equal?(left, right) do
+      flunk("""
+      Tensor assertion failed.
+      left: #{inspect(left)}
+      right: #{inspect(right)}
+      """)
+    end
+  end
+
+  def tensor_equal?(left, right) do
```

Comment:
Should this be private?

Reply:
Yes. It was originally part of one of the assert_equal clauses but we wanted to reuse it in both.

Forgot to set it as defp :p